src/rdkafka_cgrp.c

```diff
@@ -3056,17 +3862,78 @@ static void rd_kafka_cgrp_rebalance (rd_kafka_cgrp_t *rkcg,
 
         rd_atomic32_set(&rkcg->rkcg_assignment_lost, assignment_lost);
 
-        /* Remove assignment (async), if any. If there is already an
-         * unassign in progress we don't need to bother. */
-        if (!RD_KAFKA_CGRP_WAIT_REBALANCE_CB(rkcg) &&
-            !(rkcg->rkcg_flags & RD_KAFKA_CGRP_F_WAIT_UNASSIGN)) {
-                rkcg->rkcg_flags |= RD_KAFKA_CGRP_F_WAIT_UNASSIGN;
-
+        if (!is_cooperative) {
+                /* EAGER case - revoke partitions (which will be followed by
+                 * rejoin, if required (it is under normal operation). */
                 rd_kafka_rebalance_op(
                         rkcg,
                         RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS,
                         rkcg->rkcg_assignment, reason);
+                return;
+        }
+
+
+        /* COOPERATIVE case. Under usual circumstances, all partitions are not
+         * revoked unless terminating, or partitions are lost. */
+        rd_assert(terminating || assignment_lost);
+
+        if (rkcg->rkcg_assignment &&
+            rkcg->rkcg_assignment->cnt > 0) {
+
+                rd_kafka_dbg(rkcg->rkcg_rk,
+                        CONSUMER|RD_KAFKA_DBG_CGRP,
+                        "REBALANCE", "Group \"%.*s\" revoking "
+                        "all %d partition(s) (%s)",
+                        RD_KAFKAP_STR_PR(rkcg->rkcg_group_id),
+                        rkcg->rkcg_assignment->cnt,
+                        terminating ? "terminating"
+                        : "assignment lost");
+
+                // MH: Think through if this can ever happen.
+                rd_assert(rkcg->rkcg_rebalance_incr_assignment == NULL);
+
+                // MH: is _private here a concern? 
+                partitions = rd_kafka_topic_partition_list_copy(
```

Comment:
i'm thinking it would be better if `rd_kafka_cgrp_incremental_unassign` just always copied the partitions argument, even though that's not always necessary. but anyway, since the assert is there now, there's no practical difference to robustness really.

also, i had a leak related to this in `rd_kafka_cgrp_modify_subscription` ...